fdbserver/workloads/FailoverWithSSLag.actor.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+ * FailoverWithSSLag.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbserver/TesterInterface.actor.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "fdbserver/Knobs.h"
+#include "fdbserver/RecoveryState.h"
+#include "fdbserver/ServerDBInfo.h"
+#include "fdbrpc/simulator.h"
+#include "fdbrpc/SimulatorProcessInfo.h"
+#include "fdbclient/ManagementAPI.actor.h"
+#include "fdbclient/StatusClient.h"
+#include "flow/actorcompiler.h" // This must be the last include.
+
+// This actor tests failover with remote tlogs being in sync with primary but with remote storage servers lagging
+// behind the primary. Failover shouldn't complete until the remote storage servers also get in sync with the primary.
+struct FailoverWithSSLagWorkload : TestWorkload {
+	static constexpr auto NAME = "FailoverWithSSLagWorkload";
+	bool enabled;
+	double testDuration;
+	bool testSuccess;
+	std::vector<IPAddress> tlogs; // remote tlogs
+	std::vector<IPAddress> storages; // remote storages
+
+	FailoverWithSSLagWorkload(WorkloadContext const& wcx) : TestWorkload(wcx) {
+		enabled =
+		    !clientId && g_network->isSimulated(); // only do this on the "first" client, and only when in simulation
+		testDuration = getOption(options, "testDuration"_sr, 400.0);
+		testSuccess = true;
+		g_simulator->usableRegions = 2;
+	}
+
+	void disableFailureInjectionWorkloads(std::set<std::string>& out) const override { out.insert("all"); }
+
+	Future<Void> setup(Database const& cx) override { return Void(); }
+
+	Future<Void> start(Database const& cx) override {
+		if (enabled) {
+			return timeout(reportErrors(clogClient(this, cx), "FailoverWithSSLagError"), testDuration, Void());
+		}
+		return Void();
+	}
+
+	Future<bool> check(Database const& cx) override { return testSuccess; }
+
+	void getMetrics(std::vector<PerfMetric>& m) override {}
+
+	// Clog or unclog (based on argument "clog") connections between remote tlogs ("tlogs") and
+	// remote storages ("storages").
+	void clogUnclogRemoteStorages(bool clog, double seconds = 0) {
```

Comment:
clogging remote tlog and remote ss won't cause DC lag, since DC lag measures primary tlog's and remote tlog's version difference. Probably you can expose SS lag in status json so that the test can fetch it and verify the lag is happening.

Reply:
Data center version difference is the max of log and storage server version difference, so this should capture the storage server version difference. We can check the storage server version difference directly, as it is available in status json now.